### PR TITLE
Update to sleep menu 2

### LINKED ISSUE #38785: Useless before-sleep warning
# Describe the bug

tried to sleep, and got this warning:
survivormask
My char wasn't even wearing it at the time. 

As far as I know, there is no way to deactivate gas masks, survivor masks and such. So probably this one(okay, two. or 3 if filter mask is also affected) item should be excluded from before-sleep checks.

# Steps To Reproduce

1. Make a player with loaded survivor/gas mask.
2. Activate it.
3. Press '$' and see the warning.

# Expected behavior

No warning because it doesn't serve any purpose.

# Versions and configuration

    OS: Windows 7x64
    Game Version: 0.D-13133-g1eb14b4 [64-bit]
    Graphics Version: Tiles
    Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Modular Turrets [modular_turrets],
    Salvaged Robots [Salvaged_Robots],
    Aftershock [aftershock],
    Alternative Map Key [alt_map_key],
    sees-player icon, retrodays [sees_player_retro],
    Stats Through Skills [StatsThroughSkills],
    Magiclysm [magiclysm]
    ]

### PR DESCRIPTION
**Pull request by Pupsi-Mupsi**
#### Summary

`SUMMARY: None`

#### Purpose of change

- As @akirashirosawa suggested in #38722, the default sleep option is now only **NO** if there is an active element in the player inventory.
- And I've also changed the phrase to match cigars and the like because you can't **turn off** cigars.
- Gas masks, musical instruments and glowsticks are ignored.
- New FLAG **SLEEP IGNORE** added . Items can now be easily excluded.
- New hint "You may want to extinguish or turn off:" (Not shown in screenshots.)

This fixes #38785.

Hope you like it.
<br />
#### Screenshot(s)

grafik

grafik

grafik

### CODE REVIEW

File: src/handle_action.cpp [hunk 1]
```diff
@@ -955,16 +943,38 @@ static void sleep()
         }
     }
 
-    // ask for deactivation
+    uilist as_m;
+    as_m.text = _( "<color_white>Are you sure you want to sleep?</color>" );
     std::stringstream data;
     if( !active.empty() ) {
+    // (Y)es/(S)ave before sleeping/(N)o
```
Comment (akirashirosawa) [inline src/handle_action.cpp]:
```suggestion
        // (N)o/(S)ave before sleeping/(Y)es
```

File: src/handle_action.cpp [hunk 2]
```diff
@@ -955,16 +943,38 @@ static void sleep()
...
+        // ask for deactivation
         data << as_m.text << std::endl;
-        data << _( "You may want to turn off:" ) << std::endl;
+        data << _( "You may want to take care of:" ) << std::endl;
         data << " " << std::endl;
         for( auto &a : active ) {
             data << "<color_red>" << a << "</color>" << std::endl;
         }
         as_m.text = data.str();
+    } else {
```
Comment (akirashirosawa) [inline src/handle_action.cpp]:
```suggestion
    } else {
        // (Y)es/(S)ave before sleeping/(N)o
```

**Comment by BevapDin** [comment]
Instead of moving the code blocks around (that add the menu entries), you can set the initially selected item of the menu via `uilist::selected`. That way you don't have to duplicate all those entries and the menu has a consistent appearance (entries always in the same order).

File: src/handle_action.cpp (hunk 2 above)
Comment (akirashirosawa) [reply src/handle_action.cpp]: For the first time I am doing a review, I did not know how to add a line without deleting))

**Comment by Dacendeth** [comment]
Can you fix it flaring up with gas masks? you cant actually turn those off either.

**Comment by Pupsi-Mupsi** [comment]
> 
> 
> Instead of moving the code blocks around (that add the menu entries), you can set the initially selected item of the menu via `uilist::selected`. That way you don't have to duplicate all those entries and the menu has a consistent appearance (entries always in the same order).

Sounds good. Can you help me with that?

**Comment by Dacendeth** [comment]
This also just makes where yes is inconsistent, did we really need to babyproof falling asleep with a flashlight on in the first place?

**Comment by Pupsi-Mupsi** [comment]
> 
> 
> This also just makes where yes is inconsistent, did we really need to babyproof falling asleep with a flashlight on in the first place?

I feel like it might help new players. Don't you?

**Comment by akirashirosawa** [comment]
> > Instead of moving the code blocks around (that add the menu entries), you can set the initially selected item of the menu via `uilist::selected`. That way you don't have to duplicate all those entries and the menu has a consistent appearance (entries always in the same order).
> 
> Sounds good. Can you help me with that?



```
diff --git a/src/handle_action.cpp b/src/handle_action.cpp
--- a/src/handle_action.cpp
+++ b/src/handle_action.cpp
@@ -903,15 +903,16 @@ static void sleep()
     uilist as_m;
     as_m.text = _( "<color_white>Are you sure you want to sleep?</color>" );
     // (Y)es/(S)ave before sleeping/(N)o
-    as_m.entries.emplace_back( 2, true,
-                               get_option<bool>( "FORCE_CAPITAL_YN" ) ? 'N' : 'n',
-                               _( "No." ) );
-    as_m.entries.emplace_back( 1, g->get_moves_since_last_save(),
-                               get_option<bool>( "FORCE_CAPITAL_YN" ) ? 'S' : 's',
-                               _( "Yes, and save game before sleeping." ) );
     as_m.entries.emplace_back( 0, true,
                                get_option<bool>( "FORCE_CAPITAL_YN" ) ? 'Y' : 'y',
                                _( "Yes." ) );
+    as_m.entries.emplace_back( 1, g->get_moves_since_last_save(),
+                               get_option<bool>( "FORCE_CAPITAL_YN" ) ? 'S' : 's',
+                               _( "Yes, and save game before sleeping." ) );
+    as_m.entries.emplace_back( 2, true,
+                               get_option<bool>( "FORCE_CAPITAL_YN" ) ? 'N' : 'n',
+                               _( "No." ) );
+
 
     // List all active items, bionics or mutations so player can deactivate them
     std::vector<std::string> active;
@@ -958,8 +959,9 @@ static void sleep()
     // ask for deactivation
     std::stringstream data;
     if( !active.empty() ) {
+        as_m.selected = 2;
         data << as_m.text << std::endl;
-        data << _( "You may want to turn off:" ) << std::endl;
+        data << _( "You may want to take care of:" ) << std::endl;
         data << " " << std::endl;
         for( auto &a : active ) {
             data << "<color_red>" << a << "</color>" << std::endl;
```

**Comment by Pupsi-Mupsi** [comment]
Thank you akirashirosawa!

**Comment by akirashirosawa** [comment]
> Can you fix it flaring up with gas masks? you cant actually turn those off either.

It would be nice to filter things like gas masks, glowstick, but I don't know what `flag` we can use to do it.

**Comment by Pupsi-Mupsi** [comment]
> 
> 
> > Can you fix it flaring up with gas masks? you cant actually turn those off either.
> 
> It would be nice to filter things like gas masks, glowstick, but I don't know what `flag` we can use to do it.

Maybe -> "use_action": "GASMASK".
As for the glow stick, I still don't have a good idea.

**Comment by Dacendeth** [comment]
I feel new players should learn from mistakes instead of having their hand held.

**Comment by akirashirosawa** [comment]
> Thank you akirashirosawa!

You are welcome) I don’t know how to make such big suggestion in review, sorry, but, I just copied diff again.

**Comment by Pupsi-Mupsi** [comment]
> 
> 
> I feel new players should learn from mistakes instead of having their hand held.

I understand your point.
You can still go to sleep and ignore the new hint.
Or would you prefer to remove the changes?

**Comment by akirashirosawa** [comment]
> I feel new players should learn from mistakes instead of having their hand held.

 I feel that I went to sleep with mp3 player on like 1000 times. And I am very happy about this change! It just makes the game more convenient, not more easy.

**Comment by akirashirosawa** [comment]
> > > Can you fix it flaring up with gas masks? you cant actually turn those off either.
> > 
> > 
> > It would be nice to filter things like gas masks, glowstick, but I don't know what `flag` we can use to do it.
> 
> Maybe -> "use_action": "GASMASK".
> As for the glow stick, I still don't have a good idea.

We cannot hardcode such things. A gas mask and a glowstick are what came to mind now. There may be more items, may appear more items. Need to find or add a `flag`. But I think ignoring the warning is also a good solution too. After all, it is written, that `You may want to take care of`, not `You should`.

**Comment by Pupsi-Mupsi** [comment]
> Can you fix it flaring up with gas masks? you cant actually turn those off either.

@Dacendeth
@akirashirosawa

Just added this to ignare GASKMASK.
grafik

**Comment by Pupsi-Mupsi** [comment]
> 
> We cannot hardcode such things. A gas mask and a glowstick are what came to mind now. There may be more items, may appear more items. Need to find or add a `flag`. But I think ignoring the warning is also a good solution too. After all, it is written, that `You may want to take care of`, not `You should`.

You're right. We should consider adding a FLAG to filter those things.
I would prefer to move this to another PR and wait for more feedback on that matter.

**Comment by akirashirosawa** [comment]
For note: musical instruments not a problem to sleep too.


And don't forget run `make astyle ` before commit. Last commit have `astyle regressions found` error.

**Comment by Pupsi-Mupsi** [comment]
> 
> 
> For note: musical instruments not a problem to sleep too.
> 
> 
> And don't forget run `make astyle ` before commit. Last commit have `astyle regressions found` error.

You are right on both counts.
The game takes care of the instruments, but they shouldn't be part of the hint, so I'll exclude them too.

grafik

**Comment by akirashirosawa** [comment]
> Just added this to ignore GASKMASK.

Already created issue with gas mask 🙂
I think you can add fix #38785 to PR description

**Comment by Pupsi-Mupsi** [comment]
Travis build failed. I can't figure out why.
Can anyone help?

**Comment by anothersimulacrum** [comment]
You got a seemingly unrelated test failure.
https://travis-ci.org/github/CleverRaven/Cataclysm-DDA/jobs/662464028#L1511-L1539

**Comment by Night-Pryanik** [comment]
Maybe instead of trying to create a complete list of conditions which will cover different things (tools, musical instruments, clothing, bionic etc) we should introduce a new flag and add it to all these things?

**Comment by Pupsi-Mupsi** [comment]
> 
> 
> Maybe instead of trying to create a complete list of conditions which will cover different things (tools, musical instruments, clothing, bionic etc) we should introduce a new flag and add it to all these things?

Definetly! I'll try. Hope someone will take al look and help as needed.

File: data/json/items/tool/lighting.json [hunk 3]
```diff
@@ -271,6 +271,7 @@
     "type": "TOOL",
     "name": { "str": "glowstick" },
     "description": "This is a small blue light glowstick.  Use it to bend the plastic and break the glass cylinder inside to start the reaction.  It will produce a small amount of light.",
+    "flags": [ "SLEEP_IGNORE" ]
```
Comment (akirashirosawa) [inline data/json/items/tool/lighting.json]: Actually I think, need to add a flag to `glowstick_lit` item, not to regular `glowstick`

Comment (Pupsi-Mupsi) [reply data/json/items/tool/lighting.json]: Of course. Thank you!

**Comment by Pupsi-Mupsi** [comment]
> CleverRaven deleted a comment from Pupsi-Mupsi 18 minutes ago

What was that?

**Comment by umamaistempo** [comment]
The change excluding gas mask and musical items is neat :+1:

On the other hand, I slightly disagree with the "selected option" change and the text change:
1) After doing a certain action once or a few times, the user gets used to a certain expectation in how the options are laid out, so having this subtle change based on whether or not there is something to note feels to me a bit awkward for the UI (albeit more helpful on the common case, so IDK)
2) Even though cigars can't be turned off, the previous text was very clear that something was active and if you sleep like this you might have a problem; the new text proposal is not clear on that. Having the verb turn off IMO is okay because people expect those "generic cases" in games, but you could change to extensively describe the problems: "The following items are active, turned on or lit, you might want to shut off or shut down them as side-effects might include wasted batteries or even accidental death:" :joy:

**Comment by akirashirosawa** [comment]
> After doing a certain action once or a few times, the user gets used to a certain expectation in how the options are laid out, so having this subtle change based on whether or not there is something to note feels to me a bit awkward for the UI (albeit more helpful on the common case, so IDK)

This remark was actual for first change, which changed the order of the options. For me, this was also not the best solution. But the advantages of changing option by default exceeded the minuses of changing their order.
grafik
In the current (second) version, the order of options does not change, only the selected option changes. And here I do not see the cons. For most players, including me, who press `$ + Enter` automatically - changing the selected option will help to not sleep with the mp3 player on.
grafik
For those who use keyboard shortcuts (`Y`, `S`, `N`), nothing has changed in both cases. For those who use ↑↓ keys, second case is much more convenient, when only the default option is changed, but not their order. In 90% of cases, they don’t even have to use arrows since the best option has already been chosen!

>people expect those "generic cases" in games

Well then, why make different monsters, different types of fuel, etc.?
`A common monster has bitten your common head. gg.`
Personally, I do not care about this cosmetic change. I am not a native English speaker, so these changes can even be lost in the translation. But I do care about this inappropriate trolling.
Speaking of translation. Russian translators do 4 times more work, since there are more plural forms in Russian. Why not just use only one general form? After all, **people expect "generic cases" in games**?

**Comment by umamaistempo** [comment]
@akirashirosawa I will completely disregard your whole paragraph on regards to the "generic cases" because:
1) You completely missed the point on my argument
2) That sarcastic and aggressive tone is not constructive in any way
3) It's funny that you are pointing my argument as "inappropriate trolling" and you yourself is mentioning this in a paragraph that has all the elements that one would consider a trolling (no constructive part, aggresive tone etc)
4) It goes against our CoC and personally I will point to you that, in case you want to eventually work on software engineering (or even if you already work on it), participating in discussions without consideration about everyone else will get you nowhere, so, please, think about what is the implications of what you say.

I will in good faith assume that the tone part was unintentional and happened solely because of linguistic differences but, if not, I request that you refrain from repeating such behaviour as it is not acceptable on an contribution-friendly environment and will only cause problems to everyone.

**Comment by umamaistempo** [comment]
I was thinking, the `SLEEP_IGNORE` flag is like a hack to avoid a certain behaviour; considering the cost of adding the flags and that usually we want flags (or any register like this) to be additive (ie: give an additional meaning to the thing and extend it's functionality), I think that it would be a good idea, instead, to use a flag to tag out whenever something might be destructive for an unconscious person. Something like `WARN_ON_SLEEP`, what do you think @Pupsi-Mupsi @Dacendeth @akirashirosawa @kevingranade (added you because you as a core committer might be more opinated on this subject)

**Comment by akirashirosawa** [comment]
@ZhilkinSerg
@Night-Pryanik
Человек выше потроллил своей гиперболизацией сообщений в игре, я ответил тем же, гиперболизировав сообщение об ударе обобщенного монстра. Он назвал это агрессивным. Хотя я просто пошутил точно так же как он сам выше. Потом он перешел на личности. Он решил, что имеет право указывать какое поведение приемлемо. Я не знаю кто это. Может кто-то известный. Но он первым начал троллить, и когда получил такой же ответ, он первым перешел на личности. А значит он не прав. Я не хочу продолжать этот разговор и тоже переходить на личности в след за ним, устраивая токсичный оффтоп, но и делать вид, что ничего не произошло, я тоже не собираюсь. Прошу как-то отреагировать на это токсичное поведение с переходом на личности и попыткой учить другого человека жизни в обсуждении PR к игре. Пишу по-русски, потому, что я не знаю как адекватно перевести на английский фразу «перешел на личности».

**Comment by Pupsi-Mupsi** [comment]
Hello everybody! I love cdda. I just wanted to contribute to the game. I believe that any discussion can produce good things, and I really appreciate every contribution. Sometimes two contradicting opinions can be right and good in their own way. I hope that if possible we can find a solution that most players like. If not, I'm fine if we get rid of the rather small changes to the sleep menu.

**Comment by Dacendeth** [comment]
I like the bright red text, I dont like yes not being on the top when I want to sleep as it screws with years of muscle memory. I wouldnt mind it warning me about things like the gas mask if yes is on top always.

**Comment by Pupsi-Mupsi** [comment]
> 
> 
> I like the bright red text, I dont like yes not being on the top when I want to sleep as it screws with years of muscle memory. I wouldnt mind it warning me about things like the gas mask if yes is on top always.

Thanks!

Is there an easy way to ask players what they like?
Some kind of straw poll?

**Comment by akirashirosawa** [comment]
> Hello everybody! I love cdda. I just wanted to contribute to the game.

So do I. And I don't understand why someone can afford to talk about me as a person and teach me how to live, when we talk about the game.

**Comment by akirashirosawa** [comment]
> Is there an easy way to ask players what they like?
> Some kind of straw poll?

I think the emoji  👍and 👎are used for this, under your first message. So far, 8 👍0 👎

**Comment by akirashirosawa** [comment]
> I like the bright red text, I dont like yes not being on the top when I want to sleep as it screws with years of muscle memory. I wouldnt mind it warning me about things like the gas mask if yes is on top always. I cant think of any other window where yes and no swap based on something, so it doing it here feels very out of place.

I also do not like swap options. As I wrote above:

> In the current (second) version, the order of options does not change, only the selected option changes. And here I do not see the cons. For most players, including me, who press `$ + Enter` automatically - changing the selected option will help to not sleep with the mp3 player on.
> grafik
> For those who use keyboard shortcuts (`Y`, `S`, `N`), nothing has changed in both cases. For those who use ↑↓ keys, second case is much more convenient, when only the default option is changed, but not their order. In 90% of cases, they don’t even have to use arrows since the best option has already been chosen!

How do you feel about this version?

**Comment by Night-Pryanik** [comment]
@akirashirosawa честно говоря, я не нашёл чего-то прям явно "тролльского" в том посте. Человек гиперболизировал, ну так это же не преступление. Он даже смайл поставил. Не думаю, что он имел какие-то дурные намерения.
Предлагаю не усугублять ситуацию, не ругаться и не ссориться. Обсуждать изменения этого пиара в конструктивном ключе, не переходить на личности (даже в ответ на переход на личности), не вестись на провокации и самому не провоцировать, по умолчанию предполагать добрые намерения.

**Comment by Pupsi-Mupsi** [comment]
So, should I change anything else?
Or are we good?

I will post three more comments. Please vote with 👍and 👎. So we could find a way to deal with it. By the way - I will vote well. OK?

And please let me know if I missed anything else.

**Comment by Pupsi-Mupsi** [comment]
I could offer to remove the whole sentence ("You may want to take care of:").
There is already a question of whether you are really sure you want to sleep, followed by a list of items.

**Comment by Pupsi-Mupsi** [comment]
I could create a "WARN_ON_SLEEP" flag and mark all the items that should be on the waring list.

**Comment by Pupsi-Mupsi** [comment]
The third is what should happen with the option order. Leave it as is (this PR) or change it?

grafik

**Comment by umamaistempo** [comment]
About the flag, I would say that even if we consider doing it, it does not have to be shipped along with this PR, so don't worry.

About the message, having _any_ message is better than none; so between removing it and keeping it, keeping seems to me like the safest choice.

I think that we could have a more explicit message on what are the problems and solutions (ie: "turn off electronics to avoid wasting power and extinguish fires to avoid ... fires ?").

**Comment by Pupsi-Mupsi** [comment]
Thanks.

**Comment by Pupsi-Mupsi** [comment]
Need help resolving _conflicting files src/cata_string_consts.h_.

Is it enough to move
`static const std::string flag_SLEEP_IGNORE( "SLEEP_IGNORE" );`

to

src\handle_action.cpp

cause there is no "src/cata_string_consts.h" anymore.

**Comment by anothersimulacrum** [comment]
Yes.

**Pull request merged**